gnome: Package all the Gnome extensions … again?!

piegamesde opened the pull request:
![image](https://user-images.githubusercontent.com/131599/109440092-4d12dc80-7a31-11eb-88c3-b4425f538b21.png)

Take II, aka "Revert #118184" CC @mweinelt @maralorn

###### Motivation for this change

An automatic way to do this that scales up and requires little manual intervention is really needed. It works by scraping extensions.gnome.org with a python script,
that writes all relevant information into the `extensions.json`. Every attribute of besaid file can be built into a package using `buildShellExtension`.

Extensions are grouped by Gnome shell version for practical reasons. Only extensions for Gnome 3.36 and 3.38 were added for now. If desired, we can always add older versions later on.
The extensions are exposed as an attrset, `pkgs.gnome36Extensions` and `pkgs.gnome38Extensions` respectively. The extension's UUID is used as attribute name.

`pkgs.gnomeExtensions` is now enhanced with automatically generated extensions, taken from the set matching the currently packaged Gnome Shell. It is manually curated and all attributes have a proper human-friendly name. Manual conflict resolution has to be done when there are multiple extensions that share the same human-friendly name. The previously packaged extensions are left as-is, but in the future we can remove those who don't require manual packaging in the future.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review pr 108414"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

r-rmcgibbo commented:
Result of `nixpkgs-review pr 118232` at 8c60f059 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>83 packages failed to build:</summary>
  <ul>
    <li><a href="https://gist.github.com/bddcd3588ac3ca701893df0895494519">gnomeExtensions.activities-configurator</a></li>
    <li><a href="https://gist.github.com/abeec57fa39228af75003e23fb37257e">gnomeExtensions.alttab-mod</a></li>
    <li><a href="https://gist.github.com/f4c76679e2b9d9a73a0ef8e21f9013f1">gnomeExtensions.appindicator-support</a></li>
    <li><a href="https://gist.github.com/63d099fa4fdea471734c4c045d05becb">gnomeExtensions.applications-menu</a></li>
    <li><a href="https://gist.github.com/de7f05d1b2ed12b2c5027cbc2f06b175">gnomeExtensions.applications-overview-tooltip</a></li>
    <li><a href="https://gist.github.com/c013bd689d4c908f8178597450529698">gnomeExtensions.archlinux-updates-indicator</a></li>
    <li><a href="https://gist.github.com/2562ebd610fd1808edc0b371d7e56b14">gnomeExtensions.arrange-windows</a></li>
    <li><a href="https://gist.github.com/58aeb9b881e46b7723c287a06c60e460">gnomeExtensions.auto-move-windows</a></li>
    <li><a href="https://gist.github.com/02e2e2d7a9f93fbf41c079560a075c06">gnomeExtensions.babar</a></li>
    <li><a href="https://gist.github.com/517337c0061731408214cccb6b52eb67">gnomeExtensions.bing-wallpaper-changer</a></li>
    <li><a href="https://gist.github.com/3e0a7666bb054679ffb8cc31a3f9d6e2">gnomeExtensions.bluetooth-quick-connect</a></li>
    <li><a href="https://gist.github.com/1f12cc5bd5af7c0c0c5d8a24271178f1">gnomeExtensions.bring-out-submenu-of-power-offlogout-button</a></li>
    <li><a href="https://gist.github.com/48082eb6852466a42176bbf98aa4c991">gnomeExtensions.color-picker</a></li>
    <li><a href="https://gist.github.com/d8717663af06a727d0518ac5b1bc381d">gnomeExtensions.compiz-alike-magic-lamp-effect</a></li>
    <li><a href="https://gist.github.com/a54dc12b3be33f4b349703ce9a858989">gnomeExtensions.compiz-alike-windows-effect</a></li>
    <li><a href="https://gist.github.com/7584b08d4a715b2fdcd4a2499b4966c6">gnomeExtensions.compiz-windows-effect</a></li>
    <li><a href="https://gist.github.com/0142d3cc5549f153ffb0c370087d8adf">gnomeExtensions.control-blur-effect-on-lock-screen</a></li>
    <li><a href="https://gist.github.com/6e663583add84ec01022cc0b01a2b4e4">gnomeExtensions.custom-vpn-toggler</a></li>
    <li><a href="https://gist.github.com/685173d468a1a684ef0675b3aad273c2">gnomeExtensions.ddterm</a></li>
    <li><a href="https://gist.github.com/13ee7f765ae53e6c07b63ddb8c4fb461">gnomeExtensions.desktop-lyric</a></li>
    <li><a href="https://gist.github.com/4e789fcb93b5dad5092171b9af2ee5c2">gnomeExtensions.disconnect-wifi</a></li>
    <li><a href="https://gist.github.com/a6cb5b10299c5652006fab33a60b8161">gnomeExtensions.eruption-profile-switcher</a></li>
    <li><a href="https://gist.github.com/bcc3bba2e48e5d0a6ce07a911dcafaa2">gnomeExtensions.extension-list</a></li>
    <li><a href="https://gist.github.com/13e6d25185dbab7d6c86426458af7140">gnomeExtensions.extensions-in-system-menu</a></li>
    <li><a href="https://gist.github.com/a805868860debbdd813b156b41a69c31">gnomeExtensions.frippery-applications-menu</a></li>
    <li><a href="https://gist.github.com/dac30e337db6c0ede4a88693f331014f">gnomeExtensions.fullscreen-hot-corner</a></li>
    <li><a href="https://gist.github.com/380708412988da0185f4bd204dcf5ff1">gnomeExtensions.fullscreen-notifications</a></li>
    <li><a href="https://gist.github.com/558d952c4dc751994d4b56f712ddd271">gnomeExtensions.haguichi-indicator</a></li>
    <li><a href="https://gist.github.com/9cd1eed597c235ee2eacd4f410403921">gnomeExtensions.hue-lights</a></li>
    <li><a href="https://gist.github.com/32c03ef84af438ad7bb3069b7a8a48a5">gnomeExtensions.ibus-tweaker</a></li>
    <li><a href="https://gist.github.com/159b2ea682d6216d89f49a29929fae71">gnomeExtensions.ideapad</a></li>
    <li><a href="https://gist.github.com/fe5d0b63420c2bfe677b939a634efac2">gnomeExtensions.just-perfection</a></li>
    <li><a href="https://gist.github.com/5b199e176cd8630caad39d03d262e020">gnomeExtensions.kimpanel</a></li>
    <li><a href="https://gist.github.com/033eec6418dac5c73f9beb073a7ffb66">gnomeExtensions.kitchen-timer</a></li>
    <li><a href="https://gist.github.com/6593ebd1df9b6b3f88d5afda95e881d2">gnomeExtensions.latency-monitor</a></li>
    <li><a href="https://gist.github.com/a469dcd7f45af4e7ae8b849fad13e231">gnomeExtensions.launch-new-instance</a></li>
    <li><a href="https://gist.github.com/7cd5acf3653fc3d5c2f4ba9a622b5e1a">gnomeExtensions.light-dict</a></li>
    <li><a href="https://gist.github.com/031c341018a34ea450d2c24dd70955b5">gnomeExtensions.lock-keys</a></li>
    <li><a href="https://gist.github.com/dd15b25cee8ae4a6bf1b02ca1d890606">gnomeExtensions.lock-keys-2</a></li>
    <li><a href="https://gist.github.com/60017d600d36ab3a22b0ec5652fac304">gnomeExtensions.messaging-menu</a></li>
    <li><a href="https://gist.github.com/fb9f39316c957c09e240ffe38c827680">gnomeExtensions.move-clock</a></li>
    <li><a href="https://gist.github.com/646db59beb5863a7d6dd27dacee8cf34">gnomeExtensions.native-window-placement</a></li>
    <li><a href="https://gist.github.com/8223fee6a9694f55ae0acb99323e0850">gnomeExtensions.net-speed-simplified</a></li>
    <li><a href="https://gist.github.com/f5e73366278d7f1fc20543a6a246202a">gnomeExtensions.new-mail-indicator</a></li>
    <li><a href="https://gist.github.com/2b9e12a2731e6b473f2f17b98811ac21">gnomeExtensions.night-light-slider</a></li>
    <li><a href="https://gist.github.com/ca0dc359e8edc2e940024ecf70ba6fc6">gnomeExtensions.overview-improved</a></li>
    <li><a href="https://gist.github.com/49db3cebd30540015f06427348536a17">gnomeExtensions.panel-favorites</a></li>
    <li><a href="https://gist.github.com/f95fccd63657ac60d2fec5d24d1f8c15">gnomeExtensions.panel-scroll</a></li>
    <li><a href="https://gist.github.com/32b64bad7c0b37b9695a21342ace248a">gnomeExtensions.persian-calendar</a></li>
    <li><a href="https://gist.github.com/c1bf9a63b8b9a46dffb5bc3737199831">gnomeExtensions.pixel-saver</a></li>
    <li><a href="https://gist.github.com/525fbd0f7a3095646d0d91b8a96f8ee2">gnomeExtensions.places-status-indicator</a></li>
    <li><a href="https://gist.github.com/02f7d59379b4396ddd3f55e7c77696d4">gnomeExtensions.random-wallpaper</a></li>
    <li><a href="https://gist.github.com/8d86a1669ab9e8d57274f093ed379e67">gnomeExtensions.removable-drive-menu</a></li>
    <li><a href="https://gist.github.com/f7b3e79aac8715bbb3cb8d5db3f3143c">gnomeExtensions.run-or-raise</a></li>
    <li><a href="https://gist.github.com/2623a0ec34b72fa661090cdccbe95a61">gnomeExtensions.screen-word-translate</a></li>
    <li><a href="https://gist.github.com/75e8323d01bd72d0d9e8374c5669b752">gnomeExtensions.screenshot-window-sizer</a></li>
    <li><a href="https://gist.github.com/1ab7d19d52d26052e8bdb2af3eada6c8">gnomeExtensions.settingscenter</a></li>
    <li><a href="https://gist.github.com/6b283be659c353e8ae56688587260d7b">gnomeExtensions.shu-zhi</a></li>
    <li><a href="https://gist.github.com/ce0be3d1e50783d8f954b4bfcbaff75f">gnomeExtensions.simple-net-speed</a></li>
    <li><a href="https://gist.github.com/9ead6f8184bb5107ba117d5f1b7425b9">gnomeExtensions.snowy</a></li>
    <li><a href="https://gist.github.com/24b1de4759448c9b91df5e1b0a948fe2">gnomeExtensions.soft-brightness</a></li>
    <li><a href="https://gist.github.com/614b2c6c206905001b9ead6abb38d0ce">gnomeExtensions.ss-subscriber</a></li>
    <li><a href="https://gist.github.com/a853f59eb67883f8b6d33c68324d26af">gnomeExtensions.ssh-search-provider-reborn</a></li>
    <li><a href="https://gist.github.com/086e9118c33c24a8415c235474e44f28">gnomeExtensions.switch-workspace</a></li>
    <li><a href="https://gist.github.com/99cf3d9e779adccd555f3e8e8c908eb8">gnomeExtensions.syncthing-indicator</a></li>
    <li><a href="https://gist.github.com/4b4b657447f5bfc8f60785fdeee18e0b">gnomeExtensions.system-action-hibernate</a></li>
    <li><a href="https://gist.github.com/0ed5a979df6df8b6bf7bf0d2b9fe0aa3">gnomeExtensions.task-widget</a></li>
    <li><a href="https://gist.github.com/2b1c1106c5c27f60f5b45e2a98258c55">gnomeExtensions.tiling-assistant</a></li>
    <li><a href="https://gist.github.com/99777679b93f7b94d98d4e785a1b1f74">gnomeExtensions.todotxt</a></li>
    <li><a href="https://gist.github.com/1b8c41ceb82a5e6f725356710174cf52">gnomeExtensions.transparent-window</a></li>
    <li><a href="https://gist.github.com/c2c0a5dba4cca118093b1e33f4848184">gnomeExtensions.tweaks-in-system-menu</a></li>
    <li><a href="https://gist.github.com/794b9073a2c176a8ca3bb30fb980fc6d">gnomeExtensions.unblank</a></li>
    <li><a href="https://gist.github.com/204a531d0c8a868181aa0898d0708164">gnomeExtensions.unlock-dialog-background</a></li>
    <li><a href="https://gist.github.com/2913536031df8539a6ffb9015b87f2c2">gnomeExtensions.user-themes</a></li>
    <li><a href="https://gist.github.com/766d852509553a0280e4ad17d6d8730e">gnomeExtensions.user-themes-x</a></li>
    <li><a href="https://gist.github.com/5004c8d875a295a697c0ff7bb30ebf74">gnomeExtensions.utcclock</a></li>
    <li><a href="https://gist.github.com/15ac01a3093e94a72fdbbdc29ba3ff73">gnomeExtensions.wandering-pixel</a></li>
    <li><a href="https://gist.github.com/bfc8d05383524155149ddfb916f1dcdf">gnomeExtensions.weeks-start-on-monday-again</a></li>
    <li><a href="https://gist.github.com/1745a7f4957b10fe63bd9ec054a06021">gnomeExtensions.window-list</a></li>
    <li><a href="https://gist.github.com/07da86d99faa606899ac381139a2887e">gnomeExtensions.windownavigator</a></li>
    <li><a href="https://gist.github.com/a811f2088f262a80211858a309496ae6">gnomeExtensions.workspace-indicator</a></li>
    <li><a href="https://gist.github.com/39324d15f438339ef7bcb3a8651488cc">gnomeExtensions.workspace-indicator-2</a></li>
    <li><a href="https://gist.github.com/f2a3692527ddca17a41ad6d6500c6033">gnomeExtensions.x11-gestures</a></li>
  </ul>
</details>
<details>
  <summary>6 packages skipped due to time constraints:</summary>
  <ul>
    <li>gnomeExtensions.big-avatar</li>
    <li>gnomeExtensions.bitcoin-markets</li>
    <li>gnomeExtensions.bottom-panel</li>
    <li>gnomeExtensions.click-to-close-overview</li>
    <li>gnomeExtensions.jetbrains-search-provider</li>
    <li>gnomeExtensions.spotify-artwork-fixer</li>
  </ul>
</details>
<details>
  <summary>215 packages built successfully:</summary>
  <ul>
    <li>gnomeExtensions.activities-icons</li>
    <li>gnomeExtensions.activities-menu-for-apps-and-windows</li>
    <li>gnomeExtensions.activity-app-launcher</li>
    <li>gnomeExtensions.add-to-desktop</li>
    <li>gnomeExtensions.adwaita-theme-switcher</li>
    <li>gnomeExtensions.airpods-battery-status</li>
    <li>gnomeExtensions.all-ip-addresses</li>
    <li>gnomeExtensions.always-show-titles-in-overview</li>
    <li>gnomeExtensions.animation-tweaks</li>
    <li>gnomeExtensions.app-grid-tweaks</li>
    <li>gnomeExtensions.app-switcher-current-workspace-first</li>
    <li>gnomeExtensions.application-view-text</li>
    <li>gnomeExtensions.application-volume-mixer</li>
    <li>gnomeExtensions.arc-menu</li>
    <li>gnomeExtensions.ascii-emoji-buckets</li>
    <li>gnomeExtensions.audio-output-switcher</li>
    <li>gnomeExtensions.auto-select-headset</li>
    <li>gnomeExtensions.azan-islamic-prayer-times</li>
    <li>gnomeExtensions.backslide</li>
    <li>gnomeExtensions.battery-time-percentage-compact</li>
    <li>gnomeExtensions.binaryclock</li>
    <li>gnomeExtensions.bluetooth-battery</li>
    <li>gnomeExtensions.blur-my-shell</li>
    <li>gnomeExtensions.blur-provider</li>
    <li>gnomeExtensions.bottompanel</li>
    <li>gnomeExtensions.bowser-gnome-extension</li>
    <li>gnomeExtensions.break-reminder</li>
    <li>gnomeExtensions.brightness-control-using-ddcutil</li>
    <li>gnomeExtensions.brightness-panel-menu-indicator</li>
    <li>gnomeExtensions.bubblemail</li>
    <li>gnomeExtensions.calculator-button</li>
    <li>gnomeExtensions.cleaner-overview</li>
    <li>gnomeExtensions.cmus-status</li>
    <li>gnomeExtensions.color-app-menu-icon</li>
    <li>gnomeExtensions.corona-tracker</li>
    <li>gnomeExtensions.coverflow-alt-tab</li>
    <li>gnomeExtensions.cpu-power-governor</li>
    <li>gnomeExtensions.cpu-power-manager</li>
    <li>gnomeExtensions.custom-3-finger-gestures</li>
    <li>gnomeExtensions.deja-dup-backup-button</li>
    <li>gnomeExtensions.denon-avr-controler</li>
    <li>gnomeExtensions.desk-changer</li>
    <li>gnomeExtensions.desktop-icons</li>
    <li>gnomeExtensions.desktop-icons-ng-ding</li>
    <li>gnomeExtensions.dim-on-battery-power</li>
    <li>gnomeExtensions.disable-unredirect-fullscreen-windows</li>
    <li>gnomeExtensions.disable-workspace-switch-animation</li>
    <li>gnomeExtensions.dismiss-notifications-on-right-click</li>
    <li>gnomeExtensions.display-switcher</li>
    <li>gnomeExtensions.distinct-windows</li>
    <li>gnomeExtensions.downfall</li>
    <li>gnomeExtensions.draw-on-you-screen</li>
    <li>gnomeExtensions.drop-down-terminal-x</li>
    <li>gnomeExtensions.duckduckgo-search-provider</li>
    <li>gnomeExtensions.easy-docker-containers</li>
    <li>gnomeExtensions.executor</li>
    <li>gnomeExtensions.extensions-sync</li>
    <li>gnomeExtensions.eye-and-mouse-extended</li>
    <li>gnomeExtensions.favorites-only-dash</li>
    <li>gnomeExtensions.files-view</li>
    <li>gnomeExtensions.floating-dock</li>
    <li>gnomeExtensions.floating-dock-2</li>
    <li>gnomeExtensions.fly-pie</li>
    <li>gnomeExtensions.fnlock-switch-thinkpad-compact-usb-keyboard</li>
    <li>gnomeExtensions.focus</li>
    <li>gnomeExtensions.force-quit</li>
    <li>gnomeExtensions.fully-transparent-top-bar</li>
    <li>gnomeExtensions.fuzzy-clock</li>
    <li>gnomeExtensions.generic-monitor</li>
    <li>gnomeExtensions.get-out-of-the-way</li>
    <li>gnomeExtensions.github-notifications</li>
    <li>gnomeExtensions.gitlab-extension</li>
    <li>gnomeExtensions.gnome-trash</li>
    <li>gnomeExtensions.gnomesome</li>
    <li>gnomeExtensions.gnordvpn-local</li>
    <li>gnomeExtensions.gtile</li>
    <li>gnomeExtensions.guillotine</li>
    <li>gnomeExtensions.hack</li>
    <li>gnomeExtensions.harddisk-led</li>
    <li>gnomeExtensions.hibernate-status-button</li>
    <li>gnomeExtensions.hide-dash</li>
    <li>gnomeExtensions.hide-dash-forked</li>
    <li>gnomeExtensions.hide-minimized</li>
    <li>gnomeExtensions.hide-panel</li>
    <li>gnomeExtensions.hide-top-bar</li>
    <li>gnomeExtensions.horizontal-workspaces</li>
    <li>gnomeExtensions.hotel-manager</li>
    <li>gnomeExtensions.hplip-menu</li>
    <li>gnomeExtensions.ibus-font-setting</li>
    <li>gnomeExtensions.icinga-checker</li>
    <li>gnomeExtensions.improved-workspace-indicator</li>
    <li>gnomeExtensions.intel-gpu-indicator</li>
    <li>gnomeExtensions.internet-radio</li>
    <li>gnomeExtensions.internet-speed-meter</li>
    <li>gnomeExtensions.ip-finder</li>
    <li>gnomeExtensions.kaskadeur</li>
    <li>gnomeExtensions.keep-awake</li>
    <li>gnomeExtensions.keyboard-modifiers-status</li>
    <li>gnomeExtensions.keypad-tiling</li>
    <li>gnomeExtensions.kitsch</li>
    <li>gnomeExtensions.krypto</li>
    <li>gnomeExtensions.looking-glass-button</li>
    <li>gnomeExtensions.lunar-calendar</li>
    <li>gnomeExtensions.mailnag</li>
    <li>gnomeExtensions.maximize-to-empty-workspace</li>
    <li>gnomeExtensions.maximize-to-workspace-with-history</li>
    <li>gnomeExtensions.middle-click-to-close-in-overview</li>
    <li>gnomeExtensions.minimalist-calendar-2</li>
    <li>gnomeExtensions.minimalist-calendar-3</li>
    <li>gnomeExtensions.minimize-all</li>
    <li>gnomeExtensions.mmod-panel</li>
    <li>gnomeExtensions.multi-monitors-add-on</li>
    <li>gnomeExtensions.mute-spotify-ads</li>
    <li>gnomeExtensions.myhiddentopbar</li>
    <li>gnomeExtensions.network-speed</li>
    <li>gnomeExtensions.no-title-bar-forked</li>
    <li>gnomeExtensions.nordvpn-connect</li>
    <li>gnomeExtensions.notejot-button</li>
    <li>gnomeExtensions.notes</li>
    <li>gnomeExtensions.nothing-to-say</li>
    <li>gnomeExtensions.notification-centerselenium-h</li>
    <li>gnomeExtensions.notification-timeout</li>
    <li>gnomeExtensions.nvidia-gpu-stats-tool</li>
    <li>gnomeExtensions.on-screen-keyboard-button</li>
    <li>gnomeExtensions.one-drive</li>
    <li>gnomeExtensions.openweather</li>
    <li>gnomeExtensions.optimus-manager-indicator</li>
    <li>gnomeExtensions.overview-full-bright</li>
    <li>gnomeExtensions.overview-navigation</li>
    <li>gnomeExtensions.panel-icons</li>
    <li>gnomeExtensions.panel-osd</li>
    <li>gnomeExtensions.panel-world-clock-lite</li>
    <li>gnomeExtensions.persistent-email-notifications</li>
    <li>gnomeExtensions.personalize</li>
    <li>gnomeExtensions.pi-hole</li>
    <li>gnomeExtensions.ping-indicator</li>
    <li>gnomeExtensions.podman-as-docker</li>
    <li>gnomeExtensions.printers</li>
    <li>gnomeExtensions.put-windows</li>
    <li>gnomeExtensions.quake-mode</li>
    <li>gnomeExtensions.quicklists</li>
    <li>gnomeExtensions.rdesktop-launcher</li>
    <li>gnomeExtensions.refresh-wifi-connections</li>
    <li>gnomeExtensions.reminder-alarm-clock</li>
    <li>gnomeExtensions.remove-alttab-delay-v2</li>
    <li>gnomeExtensions.remove-app-menu</li>
    <li>gnomeExtensions.reorder-workspaces</li>
    <li>gnomeExtensions.resource-monitor</li>
    <li>gnomeExtensions.right-click-down</li>
    <li>gnomeExtensions.rounded-corners</li>
    <li>gnomeExtensions.runcat</li>
    <li>gnomeExtensions.scanlines-effect</li>
    <li>gnomeExtensions.screenshot-directory</li>
    <li>gnomeExtensions.screenshot-locations</li>
    <li>gnomeExtensions.screenshot-tool</li>
    <li>gnomeExtensions.sensory-perception</li>
    <li>gnomeExtensions.sermon</li>
    <li>gnomeExtensions.shamsi-calendar</li>
    <li>gnomeExtensions.shelltile</li>
    <li>gnomeExtensions.shortcuts</li>
    <li>gnomeExtensions.show-application-view-when-workspace-empty</li>
    <li>gnomeExtensions.show-applications-instead-of-overview</li>
    <li>gnomeExtensions.show-desktop-button</li>
    <li>gnomeExtensions.showtime</li>
    <li>gnomeExtensions.showtime-horizontal</li>
    <li>gnomeExtensions.simple-monitor</li>
    <li>gnomeExtensions.simple-task-bar</li>
    <li>gnomeExtensions.snap-manager</li>
    <li>gnomeExtensions.snow</li>
    <li>gnomeExtensions.status-area-horizontal-spacing</li>
    <li>gnomeExtensions.sticky-r-terminal</li>
    <li>gnomeExtensions.sticky-terminal</li>
    <li>gnomeExtensions.stocks-extension</li>
    <li>gnomeExtensions.switch-focus-type</li>
    <li>gnomeExtensions.switch-two-layouts</li>
    <li>gnomeExtensions.switcher</li>
    <li>gnomeExtensions.synaptic-button</li>
    <li>gnomeExtensions.system-menu</li>
    <li>gnomeExtensions.taskbar-updated</li>
    <li>gnomeExtensions.text-translator</li>
    <li>gnomeExtensions.the-circles-desktop-widget</li>
    <li>gnomeExtensions.three-finger-window-move</li>
    <li>gnomeExtensions.time</li>
    <li>gnomeExtensions.tocuhpad-window-switcher</li>
    <li>gnomeExtensions.toggle-alacritty</li>
    <li>gnomeExtensions.toggle-night-light</li>
    <li>gnomeExtensions.top-bar-visuals-transparent-and-blur</li>
    <li>gnomeExtensions.top-indicator-app</li>
    <li>gnomeExtensions.topiconsfix</li>
    <li>gnomeExtensions.transcodeappsearch</li>
    <li>gnomeExtensions.translate-indicator</li>
    <li>gnomeExtensions.transparent-shell</li>
    <li>gnomeExtensions.transparent-top-bar</li>
    <li>gnomeExtensions.transparent-top-bar-adjustable-transparency</li>
    <li>gnomeExtensions.transparent-top-panel</li>
    <li>gnomeExtensions.transparent-window-moving</li>
    <li>gnomeExtensions.trash</li>
    <li>gnomeExtensions.tray-icons</li>
    <li>gnomeExtensions.tray-icons-reloaded</li>
    <li>gnomeExtensions.unity-like-app-switcher</li>
    <li>gnomeExtensions.unix-timestamp-clock</li>
    <li>gnomeExtensions.vitals</li>
    <li>gnomeExtensions.volume-mixer</li>
    <li>gnomeExtensions.vpnstatus-indicator</li>
    <li>gnomeExtensions.wacom-indicator</li>
    <li>gnomeExtensions.window-is-ready-notification-remover</li>
    <li>gnomeExtensions.window-search-provider</li>
    <li>gnomeExtensions.window-session-manager</li>
    <li>gnomeExtensions.window-switcher-current-workspace-first</li>
    <li>gnomeExtensions.windowoverlay-icons</li>
    <li>gnomeExtensions.worksets</li>
    <li>gnomeExtensions.workspaces-bar</li>
    <li>gnomeExtensions.workspaces-thumbnails-applet</li>
    <li>gnomeExtensions.yaru-remix-theme-toggle</li>
    <li>gnomeExtensions.zilence</li>
  </ul>
</details>
<details>
  <summary>2 suggestions:</summary>
  <ul>
    <li>warning: unused-argument

Unused argument: `name`.
Near pkgs/desktops/gnome-3/extensions/buildGnomeExtension.nix:8:7:
```
  |
8 |     , name
  |       ^
```
</li>
    <li>warning: unused-argument

Unused argument: `pkgs`.
Near pkgs/desktops/gnome-3/extensions/buildGnomeExtension.nix:1:3:
```
  |
1 | { pkgs, lib, stdenv, fetchurl, unzip }:
  |   ^
```
</li>
  </ul>
</details>

Note that build failures [may predate](https://github.com/nix-community/hydra-check) this PR, and could be nondeterministic or hardware dependent.
Please exercise your independent judgement.

piegamesde commented:
Can somebody please validate the results from the bot? It shows some hash mismatches, but works for me …

mweinelt commented:
Most errors due to hash mismatches AFAICT.

Result of `nixpkgs-review pr 118232` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>90 packages failed to build:</summary>
  <ul>
    <li>gnomeExtensions.activities-configurator</li>
    <li>gnomeExtensions.alttab-mod</li>
    <li>gnomeExtensions.appindicator-support</li>
    <li>gnomeExtensions.applications-menu</li>
    <li>gnomeExtensions.applications-overview-tooltip</li>
    <li>gnomeExtensions.archlinux-updates-indicator</li>
    <li>gnomeExtensions.arrange-windows</li>
    <li>gnomeExtensions.auto-move-windows</li>
    <li>gnomeExtensions.babar</li>
    <li>gnomeExtensions.big-avatar</li>
    <li>gnomeExtensions.bing-wallpaper-changer</li>
    <li>gnomeExtensions.bitcoin-markets</li>
    <li>gnomeExtensions.bluetooth-quick-connect</li>
    <li>gnomeExtensions.bottom-panel</li>
    <li>gnomeExtensions.bring-out-submenu-of-power-offlogout-button</li>
    <li>gnomeExtensions.color-picker</li>
    <li>gnomeExtensions.compiz-alike-magic-lamp-effect</li>
    <li>gnomeExtensions.compiz-alike-windows-effect</li>
    <li>gnomeExtensions.compiz-windows-effect</li>
    <li>gnomeExtensions.control-blur-effect-on-lock-screen</li>
    <li>gnomeExtensions.custom-vpn-toggler</li>
    <li>gnomeExtensions.ddterm</li>
    <li>gnomeExtensions.desktop-lyric</li>
    <li>gnomeExtensions.disconnect-wifi</li>
    <li>gnomeExtensions.eruption-profile-switcher</li>
    <li>gnomeExtensions.extension-list</li>
    <li>gnomeExtensions.extensions-in-system-menu</li>
    <li>gnomeExtensions.frippery-applications-menu</li>
    <li>gnomeExtensions.fullscreen-hot-corner</li>
    <li>gnomeExtensions.fullscreen-notifications</li>
    <li>gnomeExtensions.haguichi-indicator</li>
    <li>gnomeExtensions.hide-top-bar</li>
    <li>gnomeExtensions.hue-lights</li>
    <li>gnomeExtensions.ibus-tweaker</li>
    <li>gnomeExtensions.ideapad</li>
    <li>gnomeExtensions.just-perfection</li>
    <li>gnomeExtensions.kimpanel</li>
    <li>gnomeExtensions.kitchen-timer</li>
    <li>gnomeExtensions.latency-monitor</li>
    <li>gnomeExtensions.launch-new-instance</li>
    <li>gnomeExtensions.light-dict</li>
    <li>gnomeExtensions.lock-keys</li>
    <li>gnomeExtensions.lock-keys-2</li>
    <li>gnomeExtensions.maximize-to-empty-workspace</li>
    <li>gnomeExtensions.messaging-menu</li>
    <li>gnomeExtensions.move-clock</li>
    <li>gnomeExtensions.native-window-placement</li>
    <li>gnomeExtensions.net-speed-simplified</li>
    <li>gnomeExtensions.new-mail-indicator</li>
    <li>gnomeExtensions.night-light-slider</li>
    <li>gnomeExtensions.overview-improved</li>
    <li>gnomeExtensions.panel-favorites</li>
    <li>gnomeExtensions.panel-scroll</li>
    <li>gnomeExtensions.persian-calendar</li>
    <li>gnomeExtensions.pixel-saver</li>
    <li>gnomeExtensions.places-status-indicator</li>
    <li>gnomeExtensions.random-wallpaper</li>
    <li>gnomeExtensions.removable-drive-menu</li>
    <li>gnomeExtensions.resource-monitor</li>
    <li>gnomeExtensions.run-or-raise</li>
    <li>gnomeExtensions.screen-word-translate</li>
    <li>gnomeExtensions.screenshot-window-sizer</li>
    <li>gnomeExtensions.settingscenter</li>
    <li>gnomeExtensions.shu-zhi</li>
    <li>gnomeExtensions.simple-net-speed</li>
    <li>gnomeExtensions.snowy</li>
    <li>gnomeExtensions.soft-brightness</li>
    <li>gnomeExtensions.ss-subscriber</li>
    <li>gnomeExtensions.ssh-search-provider-reborn</li>
    <li>gnomeExtensions.stocks-extension</li>
    <li>gnomeExtensions.switch-workspace</li>
    <li>gnomeExtensions.syncthing-indicator</li>
    <li>gnomeExtensions.system-action-hibernate</li>
    <li>gnomeExtensions.task-widget</li>
    <li>gnomeExtensions.tiling-assistant</li>
    <li>gnomeExtensions.todotxt</li>
    <li>gnomeExtensions.transparent-window</li>
    <li>gnomeExtensions.tweaks-in-system-menu</li>
    <li>gnomeExtensions.unblank</li>
    <li>gnomeExtensions.unlock-dialog-background</li>
    <li>gnomeExtensions.user-themes</li>
    <li>gnomeExtensions.user-themes-x</li>
    <li>gnomeExtensions.utcclock</li>
    <li>gnomeExtensions.wandering-pixel</li>
    <li>gnomeExtensions.weeks-start-on-monday-again</li>
    <li>gnomeExtensions.window-list</li>
    <li>gnomeExtensions.windownavigator</li>
    <li>gnomeExtensions.workspace-indicator</li>
    <li>gnomeExtensions.workspace-indicator-2</li>
    <li>gnomeExtensions.x11-gestures</li>
  </ul>
</details>
<details>
  <summary>214 packages built:</summary>
  <ul>
    <li>gnomeExtensions.activities-icons</li>
    <li>gnomeExtensions.activities-menu-for-apps-and-windows</li>
    <li>gnomeExtensions.activity-app-launcher</li>
    <li>gnomeExtensions.add-to-desktop</li>
    <li>gnomeExtensions.adwaita-theme-switcher</li>
    <li>gnomeExtensions.airpods-battery-status</li>
    <li>gnomeExtensions.all-ip-addresses</li>
    <li>gnomeExtensions.always-show-titles-in-overview</li>
    <li>gnomeExtensions.animation-tweaks</li>
    <li>gnomeExtensions.app-grid-tweaks</li>
    <li>gnomeExtensions.app-switcher-current-workspace-first</li>
    <li>gnomeExtensions.application-view-text</li>
    <li>gnomeExtensions.application-volume-mixer</li>
    <li>gnomeExtensions.arc-menu</li>
    <li>gnomeExtensions.ascii-emoji-buckets</li>
    <li>gnomeExtensions.audio-output-switcher</li>
    <li>gnomeExtensions.auto-select-headset</li>
    <li>gnomeExtensions.azan-islamic-prayer-times</li>
    <li>gnomeExtensions.backslide</li>
    <li>gnomeExtensions.battery-time-percentage-compact</li>
    <li>gnomeExtensions.binaryclock</li>
    <li>gnomeExtensions.bluetooth-battery</li>
    <li>gnomeExtensions.blur-my-shell</li>
    <li>gnomeExtensions.blur-provider</li>
    <li>gnomeExtensions.bottompanel</li>
    <li>gnomeExtensions.bowser-gnome-extension</li>
    <li>gnomeExtensions.break-reminder</li>
    <li>gnomeExtensions.brightness-control-using-ddcutil</li>
    <li>gnomeExtensions.brightness-panel-menu-indicator</li>
    <li>gnomeExtensions.bubblemail</li>
    <li>gnomeExtensions.calculator-button</li>
    <li>gnomeExtensions.cleaner-overview</li>
    <li>gnomeExtensions.click-to-close-overview</li>
    <li>gnomeExtensions.cmus-status</li>
    <li>gnomeExtensions.color-app-menu-icon</li>
    <li>gnomeExtensions.corona-tracker</li>
    <li>gnomeExtensions.coverflow-alt-tab</li>
    <li>gnomeExtensions.cpu-power-governor</li>
    <li>gnomeExtensions.cpu-power-manager</li>
    <li>gnomeExtensions.custom-3-finger-gestures</li>
    <li>gnomeExtensions.deja-dup-backup-button</li>
    <li>gnomeExtensions.denon-avr-controler</li>
    <li>gnomeExtensions.desk-changer</li>
    <li>gnomeExtensions.desktop-icons</li>
    <li>gnomeExtensions.desktop-icons-ng-ding</li>
    <li>gnomeExtensions.dim-on-battery-power</li>
    <li>gnomeExtensions.disable-unredirect-fullscreen-windows</li>
    <li>gnomeExtensions.disable-workspace-switch-animation</li>
    <li>gnomeExtensions.dismiss-notifications-on-right-click</li>
    <li>gnomeExtensions.display-switcher</li>
    <li>gnomeExtensions.distinct-windows</li>
    <li>gnomeExtensions.downfall</li>
    <li>gnomeExtensions.draw-on-you-screen</li>
    <li>gnomeExtensions.drop-down-terminal-x</li>
    <li>gnomeExtensions.duckduckgo-search-provider</li>
    <li>gnomeExtensions.easy-docker-containers</li>
    <li>gnomeExtensions.executor</li>
    <li>gnomeExtensions.extensions-sync</li>
    <li>gnomeExtensions.eye-and-mouse-extended</li>
    <li>gnomeExtensions.favorites-only-dash</li>
    <li>gnomeExtensions.files-view</li>
    <li>gnomeExtensions.floating-dock</li>
    <li>gnomeExtensions.floating-dock-2</li>
    <li>gnomeExtensions.fly-pie</li>
    <li>gnomeExtensions.fnlock-switch-thinkpad-compact-usb-keyboard</li>
    <li>gnomeExtensions.focus</li>
    <li>gnomeExtensions.force-quit</li>
    <li>gnomeExtensions.fully-transparent-top-bar</li>
    <li>gnomeExtensions.fuzzy-clock</li>
    <li>gnomeExtensions.generic-monitor</li>
    <li>gnomeExtensions.get-out-of-the-way</li>
    <li>gnomeExtensions.github-notifications</li>
    <li>gnomeExtensions.gitlab-extension</li>
    <li>gnomeExtensions.gnome-trash</li>
    <li>gnomeExtensions.gnomesome</li>
    <li>gnomeExtensions.gnordvpn-local</li>
    <li>gnomeExtensions.gtile</li>
    <li>gnomeExtensions.guillotine</li>
    <li>gnomeExtensions.hack</li>
    <li>gnomeExtensions.harddisk-led</li>
    <li>gnomeExtensions.hibernate-status-button</li>
    <li>gnomeExtensions.hide-dash</li>
    <li>gnomeExtensions.hide-dash-forked</li>
    <li>gnomeExtensions.hide-minimized</li>
    <li>gnomeExtensions.hide-panel</li>
    <li>gnomeExtensions.horizontal-workspaces</li>
    <li>gnomeExtensions.hotel-manager</li>
    <li>gnomeExtensions.hplip-menu</li>
    <li>gnomeExtensions.ibus-font-setting</li>
    <li>gnomeExtensions.icinga-checker</li>
    <li>gnomeExtensions.improved-workspace-indicator</li>
    <li>gnomeExtensions.intel-gpu-indicator</li>
    <li>gnomeExtensions.internet-radio</li>
    <li>gnomeExtensions.internet-speed-meter</li>
    <li>gnomeExtensions.ip-finder</li>
    <li>gnomeExtensions.jetbrains-search-provider</li>
    <li>gnomeExtensions.kaskadeur</li>
    <li>gnomeExtensions.keep-awake</li>
    <li>gnomeExtensions.keyboard-modifiers-status</li>
    <li>gnomeExtensions.keypad-tiling</li>
    <li>gnomeExtensions.kitsch</li>
    <li>gnomeExtensions.krypto</li>
    <li>gnomeExtensions.looking-glass-button</li>
    <li>gnomeExtensions.lunar-calendar</li>
    <li>gnomeExtensions.mailnag</li>
    <li>gnomeExtensions.maximize-to-workspace-with-history</li>
    <li>gnomeExtensions.middle-click-to-close-in-overview</li>
    <li>gnomeExtensions.minimalist-calendar-2</li>
    <li>gnomeExtensions.minimalist-calendar-3</li>
    <li>gnomeExtensions.minimize-all</li>
    <li>gnomeExtensions.mmod-panel</li>
    <li>gnomeExtensions.multi-monitors-add-on</li>
    <li>gnomeExtensions.mute-spotify-ads</li>
    <li>gnomeExtensions.myhiddentopbar</li>
    <li>gnomeExtensions.network-speed</li>
    <li>gnomeExtensions.no-title-bar-forked</li>
    <li>gnomeExtensions.nordvpn-connect</li>
    <li>gnomeExtensions.notejot-button</li>
    <li>gnomeExtensions.notes</li>
    <li>gnomeExtensions.nothing-to-say</li>
    <li>gnomeExtensions.notification-centerselenium-h</li>
    <li>gnomeExtensions.notification-timeout</li>
    <li>gnomeExtensions.nvidia-gpu-stats-tool</li>
    <li>gnomeExtensions.on-screen-keyboard-button</li>
    <li>gnomeExtensions.one-drive</li>
    <li>gnomeExtensions.openweather</li>
    <li>gnomeExtensions.optimus-manager-indicator</li>
    <li>gnomeExtensions.overview-full-bright</li>
    <li>gnomeExtensions.overview-navigation</li>
    <li>gnomeExtensions.panel-icons</li>
    <li>gnomeExtensions.panel-osd</li>
    <li>gnomeExtensions.panel-world-clock-lite</li>
    <li>gnomeExtensions.persistent-email-notifications</li>
    <li>gnomeExtensions.personalize</li>
    <li>gnomeExtensions.pi-hole</li>
    <li>gnomeExtensions.ping-indicator</li>
    <li>gnomeExtensions.podman-as-docker</li>
    <li>gnomeExtensions.printers</li>
    <li>gnomeExtensions.put-windows</li>
    <li>gnomeExtensions.quake-mode</li>
    <li>gnomeExtensions.quicklists</li>
    <li>gnomeExtensions.rdesktop-launcher</li>
    <li>gnomeExtensions.refresh-wifi-connections</li>
    <li>gnomeExtensions.reminder-alarm-clock</li>
    <li>gnomeExtensions.remove-alttab-delay-v2</li>
    <li>gnomeExtensions.remove-app-menu</li>
    <li>gnomeExtensions.reorder-workspaces</li>
    <li>gnomeExtensions.right-click-down</li>
    <li>gnomeExtensions.rounded-corners</li>
    <li>gnomeExtensions.runcat</li>
    <li>gnomeExtensions.scanlines-effect</li>
    <li>gnomeExtensions.screenshot-directory</li>
    <li>gnomeExtensions.screenshot-locations</li>
    <li>gnomeExtensions.screenshot-tool</li>
    <li>gnomeExtensions.sensory-perception</li>
    <li>gnomeExtensions.sermon</li>
    <li>gnomeExtensions.shamsi-calendar</li>
    <li>gnomeExtensions.shelltile</li>
    <li>gnomeExtensions.shortcuts</li>
    <li>gnomeExtensions.show-application-view-when-workspace-empty</li>
    <li>gnomeExtensions.show-applications-instead-of-overview</li>
    <li>gnomeExtensions.show-desktop-button</li>
    <li>gnomeExtensions.showtime</li>
    <li>gnomeExtensions.showtime-horizontal</li>
    <li>gnomeExtensions.simple-monitor</li>
    <li>gnomeExtensions.simple-task-bar</li>
    <li>gnomeExtensions.snap-manager</li>
    <li>gnomeExtensions.snow</li>
    <li>gnomeExtensions.spotify-artwork-fixer</li>
    <li>gnomeExtensions.status-area-horizontal-spacing</li>
    <li>gnomeExtensions.sticky-r-terminal</li>
    <li>gnomeExtensions.sticky-terminal</li>
    <li>gnomeExtensions.switch-focus-type</li>
    <li>gnomeExtensions.switch-two-layouts</li>
    <li>gnomeExtensions.switcher</li>
    <li>gnomeExtensions.synaptic-button</li>
    <li>gnomeExtensions.system-menu</li>
    <li>gnomeExtensions.taskbar-updated</li>
    <li>gnomeExtensions.text-translator</li>
    <li>gnomeExtensions.the-circles-desktop-widget</li>
    <li>gnomeExtensions.three-finger-window-move</li>
    <li>gnomeExtensions.time</li>
    <li>gnomeExtensions.tocuhpad-window-switcher</li>
    <li>gnomeExtensions.toggle-alacritty</li>
    <li>gnomeExtensions.toggle-night-light</li>
    <li>gnomeExtensions.top-bar-visuals-transparent-and-blur</li>
    <li>gnomeExtensions.top-indicator-app</li>
    <li>gnomeExtensions.topiconsfix</li>
    <li>gnomeExtensions.transcodeappsearch</li>
    <li>gnomeExtensions.translate-indicator</li>
    <li>gnomeExtensions.transparent-shell</li>
    <li>gnomeExtensions.transparent-top-bar</li>
    <li>gnomeExtensions.transparent-top-bar-adjustable-transparency</li>
    <li>gnomeExtensions.transparent-top-panel</li>
    <li>gnomeExtensions.transparent-window-moving</li>
    <li>gnomeExtensions.trash</li>
    <li>gnomeExtensions.tray-icons</li>
    <li>gnomeExtensions.tray-icons-reloaded</li>
    <li>gnomeExtensions.unity-like-app-switcher</li>
    <li>gnomeExtensions.unix-timestamp-clock</li>
    <li>gnomeExtensions.vitals</li>
    <li>gnomeExtensions.volume-mixer</li>
    <li>gnomeExtensions.vpnstatus-indicator</li>
    <li>gnomeExtensions.wacom-indicator</li>
    <li>gnomeExtensions.window-is-ready-notification-remover</li>
    <li>gnomeExtensions.window-search-provider</li>
    <li>gnomeExtensions.window-session-manager</li>
    <li>gnomeExtensions.window-switcher-current-workspace-first</li>
    <li>gnomeExtensions.windowoverlay-icons</li>
    <li>gnomeExtensions.worksets</li>
    <li>gnomeExtensions.workspaces-bar</li>
    <li>gnomeExtensions.workspaces-thumbnails-applet</li>
    <li>gnomeExtensions.yaru-remix-theme-toggle</li>
    <li>gnomeExtensions.zilence</li>
  </ul>
</details>

piegamesde commented:
Okay, I can reproduce after a garbage collect. The URLs changed their hash, wtf?

SuperSandro2000 commented:
@ofborg eval

rhoriguchi commented:
Please exclude [fuzzy-app-search](https://github.com/NixOS/nixpkgs/tree/master/pkgs/desktops/gnome-3/extensions/fuzzy-app-search), it has a patch which is needed for it to work.

piegamesde commented:
Thanks for the notice. The current approach is to package *all* of them in a hit-or-miss style (`pkgs.gnome36Extensions` and `pkgs.gnome38Extensions`), and then have a curated `pkgs.gnomeExtensions` that contains all the required patches and curation to make it work well. `pkgs.gnomeExtensions` will also be in version sync with the current `gnome-shell` version.

piegamesde commented:
I finally found the root cause for the hash mismatches. The description can be updated at any point it time through the web site, and that is reflected in the `metadata.json`.

piegamesde commented:
It's *really* ugly, but I think the hash mismatches are now finally and completely resolved.

jtojnar commented on pkgs/desktops/gnome-3/extensions/buildGnomeExtension.nix:
```diff
@@ -0,0 +1,63 @@
+      # hash.
+      extraPostFetch = let
+        # The replacement of \\u -> \u is to undo a workaround we do in update-extensions.py which does the opposite operation,
+        # because Nix can't handle unicode in JSON. The other substitutions are fallout from the previous.
```
Just a note that it should be in the next release: https://github.com/NixOS/nix/pull/4368

piegamesde replied:
@andir told me (IIRC) that this was not enough, one cannot simply depend on the currently released version do to upgrade compatibility or something like that. Especially, all the build tooling (Hydra, Ofborgs) needs to be updated or otherwise things will break ^^. But if I get a "go" on that one, I'll be happy to change it.

piegamesde committed:
gnome: Package all the GNOME extensions

An automatic way to do this that scales up and requires little manual intervention is really needed. It works by scraping extensions.gnome.org with a python script, that writes all relevant information into the `extensions.json`. Every attribute of besaid file can be built into a package using `buildShellExtension`.

Extensions are grouped by GNOME Shell version for practical reasons. Only extensions for GNOME 40 and 3.38 were added, as we don't support legacy GNOME versions.
The extensions are exposed as an attrset, `pkgs.gnome40Extensions` and `pkgs.gnome38Extensions` respectively. The package name of each extensions is generated automatically from its UUID.

The attribute `pkgs.gnomeExtensions` contains the officially packaged and supported extensions set. It contains all the automatically packaged extensions for the current GNOME Shell version, which are overwritten by manually packaged ones where needed. Unlike gnomeXYExtensions, the names are not UUIDs, but automatically generated human-friendly names. Naming collisions – which are tracked in collisions.json – need to be manually resolved in the `extensionRenames` attrset.

jtojnar commented on pkgs/desktops/gnome/extensions/default.nix:
```diff
@@ -0,0 +1,136 @@
+  };
+
+  # Take all extensions from the index that match the gnome version, build them and put them into an attrset that is ready to go into nixpkgs
+  produceExtensions = shell-version:
```
Maybe call it `produceExtensionList`.

jtojnar commented on pkgs/desktops/gnome/extensions/default.nix:
```diff
@@ -0,0 +1,136 @@
+
+  # Map the list of extensions to an attrset based on the pname as key, which is more human readable than the UUID
+  # We also take care of conflict renaming in here
+  mapReadableNames = extensions: lib.trivial.pipe extensions [
```
The argument could be called `extensionList` as well to distinguish it from the attribute set.

jtojnar commented on pkgs/desktops/gnome/extensions/default.nix:
```diff
@@ -0,0 +1,136 @@
+    # Map all extensions to their pname
+    (map (extension:
+      let
+        newName = if builtins.hasAttr extension.extensionUuid extensionRenames then
+          builtins.getAttr extension.extensionUuid extensionRenames
+        else
+          extension.pname;
```
You can use `extensionRenames.${extension.extensionUuid} or extension.pname`.

jtojnar reviewed (commented):
Reviewed the Nix part, looks great so far.

piegamesde commented on pkgs/desktops/gnome/extensions/collisions.json:
```diff
@@ -0,0 +1,74 @@
+{
```
I am committing this because the diff is really useful when doing updates. I thought about adding a test that checks for collision handling, but I personally don't need it for maintaining.

aaronjanse commented:
This is exciting! Thank you @piegamesde!

jtojnar reviewed (commented):
Just a partial review of the update script. It is quite cleanly written and documented. I had some trouble following the logic in `generate_extension_versions` but starting from the entrypoint at the bottom would probably make it clearer. Will try it when I get some more time.

I am a bit worried about the running time. How long does it usually take to finish? For example, running `maintainers/scripts/update.py` on gnome package set serially was unbearable and that is only about two hundred packages. Do you think this would be parallelizable (e.g. using asyncio like the update.nix)?

piegamesde commented on pkgs/desktops/gnome/extensions/update-extensions.py:
```diff
@@ -0,0 +1,278 @@
+    uuid = uuid.replace("@", "")
+    url: str = f"https://extensions.gnome.org/extension-data/{uuid}.v{version}.shell-extension.zip"
+
+    # Yes, we download that file three times:
```
This is mostly a historic artifact of how things went. In an ideal world, only the `nix-prefetch-url` would be needed anyways. If you don't consider this a blocker, I'd like to improve this in a second PR.

> And for maintainer, you could just log the URL.

No, I need the full zip files in order to debug potential issues.

piegamesde commented:
> It is quite cleanly written and documented.

Most of the praise goes to Andi's initial critical reviews.

> I am a bit worried about the running time.

It is on the rather slow side, but that can easily be improved in a future pull request. Fetching the zips only once should straight up give a 3x improvement, and then we can still talk about parallelizing downloads (but I think that one thread should cap out on the wandwidth, since it's all on the same server).

piegamesde commented on pkgs/desktops/gnome/extensions/update-extensions.py:
```diff
@@ -0,0 +1,278 @@
+) -> Dict[str, Dict[str, str]]:
+    """
+    Takes in a mapping from shell versions to extension versions and transforms it the way we need it:
+    - Only take one extension version per Gnome Shell major version (as per `supported_versions`)
```
Always or only in "GNOME Shell"?

jtojnar replied:
Always, unless referring to a package name (but those are all lowercase and words separated by dashes).

piegamesde commented:
But since you asked about the run time, here's the `time` output on my machine: `./update-extensions.py  96.32s user 40.46s system 8% cpu 27:09.19 total`

piegamesde commented:
@jtojnar I've had a look at `generate_extension_versions` again and I can understand your troubles with that function. The typing is rather loose and it uses Python's weird comprehension syntax. I think I managed to improve it by creating more type aliases and maybe adding more comments and example input/output data will help too.

But I won't push these changes right now, because I wouldn't want to accidentally cause a regression in the code just before the merge window closes. All changes and improvements that are not a merge blocker will have go in a future pull request.

jtojnar commented on pkgs/desktops/gnome/extensions/update-extensions.py:
```diff
@@ -0,0 +1,281 @@
+    for shell_version in shell_version_map.keys():
+        if pname in package_name_registry[shell_version]:
+            logging.warning(f"Package name '{pname}' is colliding.")
+            package_name_registry[shell_version][pname] += [uuid]
```
Not fond of globals but we can change that later.

jtojnar commented on pkgs/desktops/gnome/extensions/update-extensions.py:
```diff
@@ -0,0 +1,281 @@
+        # This allows for the diffs to be manageable (one line of change per extension) despite their quantity
+        for index, extension in enumerate(processed_extensions):
+            if index == 0:
+                out.write("[   ")
```
Why this spacing after `[` and `,`?

piegamesde replied:
To indent the following content of that line (`json.dump(extension, out)`). See the `extensions.json` for how it looks, that should hopefully make it clear.

jtojnar replied:
Saw that but it still is not clear to me why – if you indent all lines, it is as if you did not indent any.

piegamesde replied:
The idea is that the brackets `[]` are on the topmost level, without indentation. The items then are indented one level; four spaces. As per the usual style for lists without trailing commas, the commas are on the first level, and then comes the content with the indentation.

jtojnar replied:
Yeah, that is what is weird. I would not consider the items with leading commas indented, there is only one space after the comma, which is consistent with typographic rules.

piegamesde replied:
So shall I change this to a single space (`[ `, `, `)? Or how would you like to see the formatting?

jtojnar replied:
Yeah, that would probably be for the best. (Sorry for this nit picking.)

piegamesde replied:
Done.

jtojnar commented on pkgs/desktops/gnome/extensions/update-extensions.py:
```diff
@@ -0,0 +1,281 @@
+    with open("collisions.json", "w") as out:
+        # Filter out those that are not duplicates
+        for shell_version, collisions in package_name_registry.items():
+            package_name_registry[shell_version] = {
```
Ugh, mutation. I would just create a new collision registry through nested comprehensions.

jtojnar commented on pkgs/desktops/gnome/extensions/manuallyPackaged.nix:
```diff
@@ -0,0 +1,43 @@
+  window-is-ready-remover = callPackage ./window-is-ready-remover { };
+  workspace-matrix = callPackage ./workspace-matrix { };
+
+  nohotcorner = throw "gnomeExtensions.nohotcorner removed since 2019-10-09: Since 3.34, it is a part of GNOME Shell configurable through GNOME Tweaks.";
```
We should probably move this next to the aliases like in `aliases.nix`.

piegamesde replied:
I've never touched `aliases.nix`. Where exactly should I put it and should it go in a separate commit?

jtojnar replied:
I meant in `pkgs/desktops/gnome/extensions/default.nix`, below `// lib.optionalAttrs (config.allowAliases or false) {`. Yes, it is unrelated change so a separate commit is good.

piegamesde replied:
Ah, I see.

jtojnar commented on pkgs/desktops/gnome/extensions/update-extensions.py:
```diff
@@ -0,0 +1,281 @@
+    # The second time, we extract the metadata.json because we need it too
+    with urllib.request.urlopen(url) as response:
+        data = zipfile.ZipFile(io.BytesIO(response.read()), 'r')
+        metadata = data.read('metadata.json').hex()
```
base64 will be much more efficient.

jtojnar commented on pkgs/desktops/gnome/extensions/update-extensions.py:
```diff
@@ -0,0 +1,281 @@
+    Replace thinkgs like "\\u0123" in strings. This is required for nix < 2.3.10, as it does not support "\\u"
+    """
+    # The encode-decode trick also produces "\\n", but Nix supports "\n".
+    return string.encode("unicode-escape").decode().replace("\\n", "\n")
```
Passing `ensure_ascii=False` to `json.dumps` might avoid the need for this.

piegamesde replied:
That's a really good point. How well does Nix cope with arbitrary Unicode points in JSON strings? I fear that something will somehow break. Especially, does `ensure_ascii=False` give us any guarantees that no `\uXXXX` will ever be used?

I can still add it while keeping the existing code if you want, it'll at least make the `extensions.json` more readable (this includes package search descriptions etc.).

jtojnar replied:
According to https://docs.python.org/3/library/json.html#character-encodings, it is garbage in, garbage out:

```Python console
>>> import json
>>> print(json.dumps(json.loads('"\uffff"'), ensure_ascii=False))
"￿"
```

jtojnar replied:
Seems to work in Nix unstable at least:

```
nix-repl> builtins.fromJSON ''"￿"''   
"￿"

nix-repl> builtins.fromJSON ''"\uFFFF"''
"￿"
```

piegamesde commented on pkgs/desktops/gnome/extensions/update-extensions.py:
```diff
@@ -0,0 +1,281 @@
+            "uuid" is an extension UUID that looks like this (most of the time): "extension-name@username.domain.tld".
+                   Don't make any assumptions on it, and treat it like an opaque string!
+            "link" follows the following schema: "/extension/$number/$string/"
+                   The number is some monotonically increasing and unique to every extension.
```
Nope, but the "some" was too much.

piegamesde commented on pkgs/desktops/gnome/extensions/update-extensions.py:
```diff
@@ -0,0 +1,281 @@
+                    …
+                },
+                "metadata": <hex>
+            }
```
I'd prefer thinking of input and output as separate. They may share a lot of the keys with the same data types, but many of the values will be modified or processed; only a few simply get passed along (in fact, only `uuid`).

jtojnar commented:
We also need to document the updating procedure (README.md with steps describing the procedure would suffice).

User docs in the newly added GNOME section would be also nice but would not block on that.

jtojnar commented on pkgs/desktops/gnome/extensions/buildGnomeExtension.nix:
```diff
@@ -0,0 +1,54 @@
+  buildGnomeExtension = {
+    # Every gnome extension has a UUID. It's the name of the extension folder once unpacked
+    # and can always be found in the metadata.json of every extension.
+      uuid
```
BTW, with indented attrs, you might want to use trailing commas: https://github.com/nix-community/nixpkgs-fmt/issues/248

piegamesde replied:
https://github.com/NixOS/nixpkgs/pull/108414#discussion_r598357209

jtojnar replied:
You would typically use `{` at the same level as the commas as currently used by nixpkgs-fmt, or indented with trailing commas as suggested in https://github.com/nix-community/nixpkgs-fmt/issues/248. As is, the style is somewhere in between.

piegamesde replied:
Better now?

jtojnar commented on pkgs/desktops/gnome/extensions/update-extensions.py:
```diff
@@ -0,0 +1,291 @@
+    for num, raw_extension in enumerate(raw_extensions):
+        processed_extension = process_extension(raw_extension)
+        if processed_extension:
+            processed_extensions += [processed_extension]
```
append

jtojnar commented on pkgs/desktops/gnome/extensions/update-extensions.py:
```diff
@@ -0,0 +1,291 @@
+            "version": str(extension_version),
+            "sha256": sha256,
+            # The downloads are impure, their metadata.json may change at any time.
+            # Thus, be back it up / pin it to remain deterministic
```
Adding a link to https://gitlab.gnome.org/Infrastructure/extensions-web/issues/137 to the comment would be helpful too.

jtojnar commented:
Thanks for bearing with me. Great work. Let me just quickly rebase it and then let’s merge it.

piegamesde committed:
gnomeExtensions: move dropped extensions next to aliases

gregg00 commented:
Thank you, much appreciated.  
I installed the NixOS w/ GUI last week, it came with Gnome 40.1.0.  Is it possible to use Gnome 3.36 extensions or Gnome 3.38 extensions with Gnome 40.1?  Are there plans to setup extensions for Gnome 40.1?

jtojnar commented:
No, GNOME 40 does not support extensions for older versions.

There already are some extensions compatible with GNOME 40 in `gnomeExtensions` attribute.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/easy-way-to-add-paths-to-gi-typelib-path/15240/2
